docs/cardano-monads.rst

```diff
@@ -0,0 +1,608 @@
+  insurmountable). Not only this is boilerplate-heavy, it's also antimodular, as
+  every effect needs to know about each other.
+
+Conclusion:
+
+* In retrospect, we could probably solve the ``N*M`` issue by careful use of
+  ``{-# OVERLAPPABLE #-}`` instances (although ``Mockable``-related type
+  families would still lead to a fairly large amount of boilerplate).
+  Nevertheless, the run-time performance of highly nested transformer stacks
+  might be unsatisfactory. (Slowdown is linear in the amount of layers).
+
+
+Using Ether
+~~~~~~~~~~~
+
+The core observation is that many domain-specific effects are isomorphic to
+general purpose ones. The majority of monad transformers in Cardano SL were
+basically clones of ``ReaderT`` or ``StateT``. This is why we decided to use
+``ether``: unlike ``mtl``, it allows using multiple transformers of the same
+sort (i.e. multiple ``ReaderT``) in the same transformer stack, so we could
+simply replace our hand-crafted transformers with those from ``ether`` and put
+an end to the ``N*M`` problem. Furthermore, ``ether`` offers *flattening*,
+solving the run-time performance issue as well.
+
+Typically, the code would look like this::
+
+    type MonadFoo = Ether.MonadReader Foo
+    type FooT = Ether.ReaderT Foo
+
+    type MonadBar = Ether.MonadReader Bar
+    type BarT = Ether.ReaderT Bar
+
+Even when we needed a custom class with methods (rather than a synonym for
+``Ether.MonadReader``), we could define a single instance for ``IdentityT``, no
+boilerplate required.
+
+We used flattening to achieve good run-time performance, but kept many custom
+classes with effect operations (to allow for varying implementations). This led
+to a peculiar situation where the transformer stack consisted mostly of
+``IdentityT``::
+
+    type M =
+        TaggedTrans FooEff IdentityT $
+        TaggedTrans BarEff IdentityT $
+        TaggedTrans BazEff IdentityT $
+        ReaderT (FooEnv, BarEnv, BazEnv) IO
+
+Pros:
+
+* Extensibility. Introducing a new effect is really cheap. The code is modular
+  and effects don't need to know about each other.
+
+* Flexibility. (Same as above, local use of effects).
+
+* Good run-time performance. Since in the end the entire monad transformer stack
+  was just ``ReaderT`` with a bunch of ``IdentityT`` on top (and occasional
+  local ``StateT``), we enjoyed good run-time performance.
+
+* Conciseness. No boilerplate.
+
+Cons:
+
+* Bad run-time configurability. (Same as above)
+
+* Bad compile-time performance. Due to the way flattening works in Ether and due
+  to a GHC bug, the compile-time performance was devastating. Turning ``-O2``
+  could mean hours of compilation and required up to 65 GIGABYTES of RAM
+  (ridiculous!). This was because GHC generated an exponential amount of
+  coercions (as evidenced by investigating .hi-files). Basically, @int-index no longer
+  can recommend Ether to people as there is no good solution to this known.
```

Comment:
Is there / are there GHC ticket(s) relevant to this? I wonder if the huge amount of memory usage is essential, or if it can be considered a bug.

Reply:
@nh2 was investigating this. I don't know if a ticket was created as the result, but I think it wasn't.

Reply:
as far as I remember you, @int-index, found a tricky workaround, when we wrap stack of transformers by newtype, don't you?